python-sdk/nuscenes/eval/lidarseg/evaluate.py

```diff
@@ -0,0 +1,208 @@
+        :param nusc: A NuScenes object.
+        :param results_folder: Path to the folder.
+        :param ignore_idx: Index of the class to be ignored in the evaluation.
+        :param verbose: Whether to print messages during the evaluation.
+        """
+        # Check there are ground truth annotations.
+        assert len(nusc.lidarseg) > 0, 'Error: No ground truth annotations found in {}.'.format(nusc.version)
+
+        # Check results folder exists.
+        self.results_folder = results_folder
+        assert os.path.exists(results_folder), 'Error: The result folder ({}) does not exist.'.format(results_folder)
+
+        self.nusc = nusc
+        self.results_folder = results_folder
+        self.ignore_idx = ignore_idx
+        self.verbose = verbose
+
+        self.global_cm = None
+
+        self.adaptor = LidarsegChallengeAdaptor(self.nusc)
+
+        self.num_classes = len(self.adaptor.merged_name_2_merged_idx_mapping)
+        if self.verbose:
+            print('There are {} classes.'.format(self.num_classes))
+
+    def evaluate(self) -> Dict:
+        """
+        Performs the actual evaluation.
+        :return: A dictionary containing the IOU of the individual classes and the mIOU.
+        """
+        for sample in tqdm(self.nusc.sample):
+            # Get the sample data token of the point cloud.
+            sd_token = sample['data']['LIDAR_TOP']
+
+            # Load the ground truth labels for the point cloud.
+            lidarseg_label_filename = os.path.join(self.nusc.dataroot,
+                                                   self.nusc.get('lidarseg', sd_token)['filename'])
+            lidarseg_label = self.load_bin_file(lidarseg_label_filename)
+
+            lidarseg_label = self.adaptor.convert_label(lidarseg_label)
+
+            # Load the predictions for the point cloud.
+            lidarseg_pred_filename = os.path.join(self.results_folder, 'lidarseg',
+                                                  self.nusc.version, sd_token + '_lidarseg.bin')
+            lidarseg_pred = self.load_bin_file(lidarseg_pred_filename)
+
+            # TODO remove!!
+            lidarseg_pred = self.adaptor.convert_label(lidarseg_pred)
+
+            # Get the confusion matrix between the ground truth and predictions.
+            sd_cm = self._get_confusion_matrix(lidarseg_label, lidarseg_pred)
+
+            # Update the confusion matrix for the sample data into the confusion matrix for the split.
+            self._update_confusion_matrix(sd_cm)
+
+        iou_per_class = self._get_per_class_iou(self.global_cm)
+        miou = np.nanmean(iou_per_class)
+
+        # Put everything nicely into a dict.
+        id2name = {idx: name for name, idx in self.adaptor.merged_name_2_merged_idx_mapping.items()}
+        results = dict()
+        for i, class_iou in enumerate(iou_per_class):
+            if not np.isnan(class_iou):
+                results['iou_' + id2name[i]] = class_iou
```

Comment:
results['iou_' + id2name[i]] = class_iou
It's a bit messy to read these out. Let's store them in
results['iou'][id2name[i]]